infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java

```diff
@@ -654,20 +655,34 @@ public void accept(ProbeResult probeResult) {
   /** Listens container's events and terminates workspace if unrecoverable event occurs. */
   public class UnrecoverableEventHanler implements ContainerEventHandler {
     private List<String> events;
+    private Map<String, Pod> workspacePods;
+    private Date handlerInitialization;
 
-    public UnrecoverableEventHanler() {
+    public UnrecoverableEventHanler(Map<String, Pod> workspacePods) {
       this.events =
           Strings.isNullOrEmpty(unrecoverableEvents)
               ? Collections.EMPTY_LIST
               : Arrays.asList(unrecoverableEvents.split(","));
+      this.workspacePods = workspacePods;
+      this.handlerInitialization = new Date();
     }
 
+    /*
+     * Event is considered to be unrecoverable if it belongs to one of the workspace pods
+     * and 'lastTimestamp' of the event is *after* the the time of handler initialization
+     */
     @Override
     public void handle(ContainerEvent event) {
-      if (isUnrecoverable(event)) {
+      String podName = event.getPodName();
+      String eventLastTimestamp = event.getLastTimestamp();
+      Date eventLastTimestampDate = convertEventTimestampToDate(eventLastTimestamp);
+
+      if (isUnrecoverable(event)
```

Comment:
I think it may sense to reorder conditions here. 
```java
 if (workspacePods.containsKey(podName) // if event is not related to workspaces pods it doesn't make sense to check other conditions
   && eventLastTimestampDate != null
   && eventLastTimestampDate.after(handlerInitialization) // if event happened before initialization it doesn't make sense to check is it recoverable or not
   && isUnrecoverable(event)) {
```
It doesn't bring so much performance benefits but still :smiley:

Reply:
Since null appears only when a date is not parsable I doubt it makes much sense to optimize )

Reply:
I'd say that main improvement (main but little one) is putting `workspacePods.containsKey(podName)` at the begging. Because containsKey in hash map is cheaper operation than sequential check list items with start with.

Reply:
good point